### PR TITLE
a better scatter for restore

### PR DESCRIPTION
**Pull request by benesch**
@danhhz this is very much WIP, so please don't review for style. The concepts are all here, though, and I know you were itching for a prototype to base our discussions on.

I just talked to Tamir, who intimated that the `wantScatter` flag on the replica will catch some flak in code review. We can make it nicer by e.g. storing the `wantScatter` state in the `replicateQueue` rather than on the `Replica` object, but his objection was to tracking the state at all, especially considering the state disappears as soon as the lease transfers and needs to be re-requested by whoever calls scatter. It is admittedly a weird API.

Do you think you and I could sit down tomorrow with someone from core just to make sure this approach will pass muster? Perhaps with @a-robinson?

The other option I see (which we talked about today) is deriving the `wantScatter` flag from properties of the range (e.g., young and empty), then teaching `RESTORE` repeatedly query those range descriptors and wait for them to evenly distribute.

### CODE REVIEW

File: pkg/ccl/sqlccl/restore.go [hunk 1]
```diff
@@ -722,20 +733,30 @@ func Restore(
...
+						if r.Error != nil {
+							log.Warningf(ctx, "error scattering range [%s,%s): %+v",
+								r.Span.Key, r.Span.EndKey, r.Error.GoError())
+							reqs = append(reqs, roachpb.AdminScatterRequest{
+								Span: r.Span,
+							})
+						}
 					}
+					time.Sleep(50 * time.Millisecond)
 				}
```
Comment (benesch) [inline pkg/ccl/sqlccl/restore.go]: A crappy implementation of repeatedly scattering all ranges until they report there's no work to be done.

File: pkg/storage/store.go [hunk 2]
```diff
@@ -432,7 +432,8 @@ type Store struct {
```
Comment (benesch) [inline pkg/storage/store.go:432]: Copy pasta of the changes to allow empty snapshots to proceed even when a big snapshot is in progress.

File: pkg/ccl/sqlccl/restore.go [hunk 3]
```diff
@@ -541,7 +545,14 @@ func presplitRanges(baseCtx context.Context, db client.DB, input []roachpb.Key)
...
 		splitKey = keys.MakeRowSentinelKey(splitKey)
-		if err := db.AdminSplit(ctx, splitKey); err != nil {
+		const maxAttempts = 3
+		if err := retry.WithMaxAttempts(ctx, base.DefaultRetryOptions(), maxAttempts, func() error {
+			err := db.AdminSplit(ctx, splitKey)
+			if err != nil {
+				log.Warningf(ctx, "failed to split range at %s: %s", splitKey, err)
+			}
+			return err
+		}); err != nil {
```
Comment (benesch) [inline pkg/ccl/sqlccl/restore.go:553]: (This change is unrelated; fixes an ambiguous result error that's been present for a while.)

File: pkg/storage/allocator.go [hunk 4]
```diff
@@ -513,7 +515,7 @@ func (a *Allocator) ShouldTransferLease(
 		log.Infof(ctx, "ShouldTransferLease (lease-holder=%d):\n%s", leaseStoreID, sl)
 	}
 
-	transferDec, _ := a.shouldTransferLeaseUsingStats(ctx, sl, source, existing, stats)
+	transferDec, _ := a.shouldTransferLeaseUsingStats(ctx, sl, source, existing, stats, wantScatter)
```
Comment (a-robinson) [inline pkg/storage/allocator.go]: Even if we do want to use `wantScatter` within `ShouldTransferLease`, we don't really need to pass `wantScatter` down - just use it to override the `shouldNotTransfer` result.

File: pkg/storage/allocator.go [hunk 5]
```diff
@@ -415,6 +415,7 @@ func (a *Allocator) TransferLeaseTarget(
 	stats *replicaStats,
 	checkTransferLeaseSource bool,
 	checkCandidateFullness bool,
+	wantScatter bool,
```
Comment (a-robinson) [inline pkg/storage/allocator.go:423]: I think you can just reuse `checkTransferLeaseSource` for this, which effectively instructs this function to not even consider keeping the lease on the local replica.

File: pkg/storage/replica_command.go [hunk 6]
```diff
@@ -3837,36 +3836,61 @@ func RelocateRange(
 func (r *Replica) adminScatter(
 	ctx context.Context, args roachpb.AdminScatterRequest,
 ) (roachpb.AdminScatterResponse, error) {
-	db := r.store.DB()
+	wantScatter := true
+	r.setWantScatter(wantScatter)
```
Comment (a-robinson) [inline pkg/storage/replica_command.go]: This needs to get unset somewhere - e.g. after replicate queue processing or after giving up the lease

File: pkg/storage/replica_command.go [hunk 7]
```diff
@@ -3837,36 +3836,61 @@ func RelocateRange(
...
+				Error: roachpb.NewError(err),
+			}},
+		}
 	}
-	for i := 0; i < num; i++ {
-		j := i + rng.Intn(len(stores)-i)
-		stores[i], stores[j] = stores[j], stores[i]
+
+	sysCfg, _ := r.store.cfg.Gossip.GetSystemConfig()
+	zone, err := sysCfg.GetZoneConfigForKey(rangeDesc.StartKey)
```
Comment (a-robinson) [inline pkg/storage/replica_command.go]: Thanks for fixing this to use the zone config

File: pkg/storage/replica_command.go [hunk 8]
```diff
@@ -3837,36 +3836,61 @@ func RelocateRange(
...
+	if err != nil {
+		return makeResponse(err), nil
 	}
-	targets := stores[:num]
 
-	relocateErr := RelocateRange(ctx, db, rangeDesc, targets)
+	lease, _ := r.getLease()
+	if !r.IsLeaseValid(lease, now) {
+		return makeResponse(errors.New("replica is not the leaseholder")), nil
```
Comment (a-robinson) [inline pkg/storage/replica_command.go]: This is more of a nit, but I'd try to use `NotLeaseholderError` here

**Review by a-robinson — COMMENTED** [review]
The mechanics of this feel pretty awkward, but I'm curious if you have any data on how it performed.
